RIOT-OS/RIOT
# cc2538_rf: implement Radio HAL

### Contribution description
This PR adds an `ieee802154_hal` implementation for the CC2538 radios. It's based on the work of #14655. It also includes the definition of an `ieee802154_radio_hal` module to be able to switch between `netdev` and `ieee802154_hal` implementations (they are mutually exclusive).

The test application was adapted from #14655 for testing this radio. When more radios are there, the test application will be expanded.




### Testing procedure

Run `tests/ieee802154_hal` test. Ideally this should be tested between 2 CC2538 radios. If that's not possible, it should be enough to use an AT86RF2XX or NRF52840 with #14655.

- Test setting channel and TX power with the `config_phy` command:
```
2020-08-12 18:24:49,390 #  config_phy
2020-08-12 18:24:49,393 # Usage: config_phy <channel> <tx_pow>
```

An out of range value for TX power should give an error

- Send data between nodes using the `txtsnd` command.
```
Usage: txtsnd <long_addr> <len>
```
It's possible to get the long address with the `print_addr` command.

- Perform CCa with the `cca` command:
```
2020-08-12 18:28:57,628 #  cca
2020-08-12 18:28:57,629 # CLEAR
> 2020-08-12 18:28:57,827 #  cca
2020-08-12 18:28:57,828 # CLEAR
> 2020-08-12 18:28:58,157 #  cca
2020-08-12 18:28:58,158 # CLEAR
> 2020-08-12 18:28:58,334 #  cca
2020-08-12 18:28:58,335 # BUSY
> 2020-08-12 18:28:58,473 #  cca
2020-08-12 18:28:58,474 # CLEAR
```

- Change the CCA mode and the threshold with the `config_cca` command:
```
config_cca <ed|cs|or|and> [ED threshold]>
```

Note the ED is in dBm. The ED threshold doesn't affect the Carrier Sense mode (CCA mode 2).

- Try different TX modes (CSMA-CA, CCA, direct transmission) using the `tx_mode` command:
```
Usage: tx_mode <csma_ca|cca|direct>
```

- Try different RX states (ACK on, ACK disabled, ACK with Frame Pending or Promiscuous mode) with the `rx_mode` command:
```
Usage: rx_mode <on|off|pend|promisc>
``` 

Use a sniffer to check if the receiver sends ACK. As expected, a radio with Promiscuous Mode will receive all kind of packets with correct FCS.

- Use the `caps` command to get the available caps for a given device:
```
2020-08-19 14:24:46,502 #  caps
2020-08-19 14:24:46,503 #                CAPS              
2020-08-19 14:24:46,505 # =================================
2020-08-19 14:24:46,507 # - Frame Retransmissions:        n
2020-08-19 14:24:46,509 #     * Info retries:             n
2020-08-19 14:24:46,511 # - Auto CSMA-CA:                 y
2020-08-19 14:24:46,513 # - 2.4 GHz band:                 y
2020-08-19 14:24:46,515 # - SubGHz band:                  n
2020-08-19 14:24:46,516 # - TX DONE indication:           y
2020-08-19 14:24:46,518 #     * ACK Timeout indication:   n
2020-08-19 14:24:46,519 # - RX_START indication:          n
2020-08-19 14:24:46,521 # - CCA Done indication:          y
```


- Use the `test_states` command to test state transitions and measure transition time. The important thing here is the TX-RX turnaround time (it should be equal or less than 12 symbols, 192 us, in order to comply with LIFS frames like the ACK):
```
2020-08-19 14:09:52,563 # test_states
2020-08-19 14:09:52,564 # Setting state to: TRX_OFF
2020-08-19 14:09:52,565 # 	Transition took 7 usecs
2020-08-19 14:09:52,566 # TRX_OFF-> RX
2020-08-19 14:09:52,578 # 	Transition took 194 usecs
2020-08-19 14:09:52,578 # RX-> TRX_OFF
2020-08-19 14:09:52,579 # 	Transition took 6 usecs
2020-08-19 14:09:52,580 # TRX_OFF-> TX
2020-08-19 14:09:52,581 # 	Transition took 7 usecs
2020-08-19 14:09:52,583 # TX-> TRX_OFF
2020-08-19 14:09:52,584 # 	Transition took 6 usecs
2020-08-19 14:09:52,584 # 
2020-08-19 14:09:52,585 # Setting state to: RX
2020-08-19 14:09:52,586 # 	Transition took 193 usecs
2020-08-19 14:09:52,587 # RX-> TRX_OFF
2020-08-19 14:09:52,588 # 	Transition took 7 usecs
2020-08-19 14:09:52,588 # TRX_OFF-> RX
2020-08-19 14:09:52,591 # 	Transition took 193 usecs
2020-08-19 14:09:52,591 # RX-> TX
2020-08-19 14:09:52,593 # 	Transition took 6 usecs
2020-08-19 14:09:52,593 # TX-> RX
2020-08-19 14:09:52,605 # 	Transition took 194 usecs
2020-08-19 14:09:52,606 # 
2020-08-19 14:09:52,607 # Setting state to: TX
2020-08-19 14:09:52,608 # 	Transition took 7 usecs
2020-08-19 14:09:52,609 # TX-> TRX_OFF
2020-08-19 14:09:52,610 # 	Transition took 6 usecs
2020-08-19 14:09:52,610 # TRX_OFF-> TX
2020-08-19 14:09:52,611 # 	Transition took 7 usecs
2020-08-19 14:09:52,612 # TX-> RX
2020-08-19 14:09:52,613 # 	Transition took 193 usecs
2020-08-19 14:09:52,613 # RX-> TX
2020-08-19 14:09:52,614 # 	Transition took 6 usecs
2020-08-19 14:09:52,614 # 
2020-08-19 14:09:52,616 # 	Transition took 193 usecs
```
Note this radio takes ~193 us to switch between RX-TX if the radio is not programmed to automatically go from TX to RX. However, since the symbol time is 16 us, this is not a problem for detecting the IEEE 802.15.4 preamble.

#14787 proves this :)

### Issues/PRs references
#13943 and friends
